### PR TITLE
Speed up find all references.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Speed up find all references by maintaining the set of all names referenced within a file. Changes `FileHash` to contain references to the names of class defs / method defs / constant literals.

### Motivation

Without this change, find all references is very slow on large codebases.

### Test plan

Automated tests exist. Even though existing "find all references" tests in test_corpus do not span multiple files, they still exercise this logic. If `setupLSPQueryBySymbol` doesn't determine that the test file has the given symbol, then it won't select that file to scan for references and the assertions in that file will fail.

### CODE REVIEW

**Comment by jvilk-stripe** [comment]
I have yet to measure how this impacts memory usage. It might be too expensive as-is. :)

File: core/NameHash.h [hunk 1]
```diff
@@ -50,7 +50,7 @@ struct UsageHash {
 
 struct FileHash {
     GlobalStateHash definitions;
-    UsageHash usages;
+    UsageHash names;
```
Comment (DarkDimius) [inline core/NameHash.h]: WDYT about keeping 2 sets of names separate?
They have semantically different behavior and you rarely want a mix of both, you want either constant names or method names.

**Comment by DarkDimius** [comment]
> It might be too expensive as-is. :)

Yeah, both of them can be converted to bloom filters but so far, due to how compact our datastructures are, I didn't feel a need :-)

File: core/NameHash.h (hunk 1 above)
Comment (jvilk-stripe) [reply core/NameHash.h]: That's fine with me!

**Comment by DarkDimius** [comment]
Super hyped about this one.
Do we believe we have enough test coverage to make sure we find if someone introduces bugs to this?
Otherwise, LGTM

**Comment by jvilk-stripe** [comment]
> Do we believe we have enough test coverage to make sure we find if someone introduces bugs to this?

No, not yet. That's part of the reason why it's WIP. After my other go-to-def/find usages PR lands, I plan to extend the test coverage to the feature to the point where this is true.

File: core/NameHash.h [hunk 2]
```diff
@@ -44,8 +44,12 @@ struct GlobalStateHash {
     UnorderedMap<NameHash, u4> methodHashes;
 };
 
+/** Sorts an array of NameHashes and removes duplicates. */
+void sortAndDedupe(std::vector<core::NameHash> &hashes);
```
Comment (DarkDimius) [inline core/NameHash.h]: we try to not declare top-level functions.
Could this be a static function in some class?

File: test/testdata/lsp/symbol_query_filter__helper.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# typed: true
```
Comment (DarkDimius) [inline test/testdata/lsp/symbol_query_filter__helper.rb]: Could we add tests for aliases?

Comment (jvilk-stripe) [reply test/testdata/lsp/symbol_query_filter__helper.rb]: @DarkDimius does this mean "Find All References" for `B` should return all references to `A` and `B`?

If so, then is there an easy way to retrieve all type aliases for a given symbol?

Comment (jvilk-stripe) [reply test/testdata/lsp/symbol_query_filter__helper.rb]: IRL; mainly, we need a test to define what behavior to use.

**Comment by jvilk-stripe** [comment]
I believe this is ready for a final review. I don't see any discernible bump in `max_rss` with this change on Stripe's codebase, so I think the increase in overhead for NameHash is fine. (I can collect some additional data if there is any concern.)

In a subsequent PR, I might parallelize indexing on the fast path to make find all references faster for symbols that are referenced in many files. Otherwise, Sorbet becomes unresponsive for longer than a slow path if the user queries one of those symbols.

File: main/lsp/lsp.h [hunk 4]
```diff
@@ -171,7 +171,8 @@ class LSPLoop {
                          const std::vector<core::FileHash> &hashes) const;
     /** Apply conservative heuristics to see if we can run a fast path, if not, bail out and run slowPath */
     TypecheckRun tryFastPath(std::unique_ptr<core::GlobalState> gs,
-                             std::vector<std::shared_ptr<core::File>> &changedFiles, bool allFiles = false);
+                             const std::vector<std::shared_ptr<core::File>> &changedFiles,
```
Comment (jvilk-stripe) [inline main/lsp/lsp.h]: I made this parameter `const`, removed the now-unused `allFiles` parameter, and added an extra parameters for files that haven't changed but should be included on the fast path (which avoids needless copying and computation on unchanged files).

File: main/lsp/requests/references.cc [hunk 5]
```diff
@@ -43,11 +43,14 @@ LSPResult LSPLoop::handleTextDocumentReferences(unique_ptr<core::GlobalState> gs
...
             } else if (auto identResp = resp->isIdent()) {
-                std::vector<std::shared_ptr<core::File>> files;
-                auto run2 = runLSPQuery(
-                    move(gs), core::lsp::Query::createVarQuery(identResp->owner, identResp->variable), files, true);
-                gs = move(run2.gs);
-                response->result = extractLocations(*gs, run2.responses);
+                auto loc = identResp->owner.data(*gs)->loc();
+                if (loc.exists()) {
+                    auto run2 =
+                        runLSPQuery(move(gs), core::lsp::Query::createVarQuery(identResp->owner, identResp->variable),
```
Comment (jvilk-stripe) [inline main/lsp/requests/references.cc]: AFAIK, references to local variables cannot span multiple files, so now I restrict local variable queries to the file that defines the symbol that owns it.

If I'm wrong about this, let me know...

File: main/lsp/updates.cc [hunk 6]
```diff
@@ -265,37 +266,29 @@ LSPLoop::TypecheckRun LSPLoop::tryFastPath(unique_ptr<core::GlobalState> gs,
...
-            changedHashes.resize(
-                std::distance(changedHashes.begin(), std::unique(changedHashes.begin(), changedHashes.end())));
+            core::NameHash::sortAndDedupe(changedHashes);
         }
     }
 
     if (takeFastPath) {
         Timer timeit(logger, "fast_path");
-        if (allFiles) {
-            subset.clear();
```
Comment (jvilk-stripe) [inline main/lsp/updates.cc]: No more `allFiles` parameter, so we have less logic here 🎉

File: main/lsp/updates.cc [hunk 7]
```diff
@@ -265,37 +266,29 @@ LSPLoop::TypecheckRun LSPLoop::tryFastPath(unique_ptr<core::GlobalState> gs,
...
+                                  oldHash.usages.sends.end(), std::back_inserter(intersection));
+            if (!intersection.empty()) {
+                auto ref = core::FileRef(i);
+                logger->debug("Added {} to update set as used a changed method",
+                              !ref.exists() ? "" : ref.data(*finalGs).path());
+                subset.emplace_back(ref);
             }
         }
+        // Remove any duplicate files.
+        fast_sort(subset);
```
Comment (jvilk-stripe) [inline main/lsp/updates.cc]: Before this dedupe logic, I think we might have been typechecking `changedFiles` twice...

File: test/helpers/position_assertions.cc [hunk 8]
```diff
@@ -505,6 +505,8 @@ void UsageAssertion::check(const UnorderedMap<string, shared_ptr<core::File>> &s
                         // Should never happen.
                         ADD_FAILURE()
                             << "Error in test runner: identical locations weren't reported as subsets of one another.";
+                        expectedLocationsIt++;
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc]: Fixes an infinite loop in the test runner that I encountered while testing this change.

File: main/lsp/requests/references.cc (hunk 5 above)
Comment (DarkDimius) [reply main/lsp/requests/references.cc]: AFAIK you're right :-)

File: main/lsp/updates.cc (hunk 7 above)
Comment (DarkDimius) [reply main/lsp/updates.cc]: Huh, are you sure? won't we report errors twice than? Why wasn't this found by tests?

Comment (jvilk-stripe) [reply main/lsp/updates.cc]: Tests suppress duplicate errors. (This was the behavior in `test_corpus` before I refactored it, and it was decided back then to keep the behavior.)

You can see the issue here:

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20%23%20Change%20'String'%20to%20'Integer'%0A%20%20sig%20%7Bparams(x%3A%20Integer).returns(String)%7D%0A%20%20def%20bar(x)%0A%20%20%20%20x.to_s%0A%20%20end%0Aend%0A%0Aj%20%3D%20T.let(A.new.bar(10)%2C%20String)%0A%0A%0A

Comment (jvilk-stripe) [reply main/lsp/updates.cc]: > won't we report errors twice then?

tl;dr: Yes; we do report errors twice.

Comment (jvilk-stripe) [reply main/lsp/updates.cc]: Duplicate error ignore logic is here:

https://github.com/sorbet/sorbet/blob/master/test/helpers/position_assertions.cc#L628
